### ISSUE TITLE
Handle status.storedVersions migration during upgrades

### ISSUE THREAD

**Issue by furkatgofurov7** [kubernetes-sigs/cluster-api#6691]
**User Story**

As a developer/user/operator, I would like to do an upgrade from old API versions to newer ones in sequence (v1a3 => v1a4 => v1beta1 ) without manually removing stored versions and applying them in the middle of the upgrade process (v1a4 => v1beta1 since v1a4 would have v1a3 `storedVersion` in the status). 

**Detailed Description**

we have a scenario where we do an upgrade from CAPM3 v1a4 => v1a5 => v1beta1 (CAPI  v1a3 => v1a4 => v1beta1). The problem is, clusterctl fails (see logs below) to upgrade from v1a5 => v1b1 because of the `status.storedVersions` in the CRD which includes `v1a4` API version in the status. 

```
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
```

 The request would be, should the clusterctl take care of removing of old version from the `status.storedVersions` or how is it handled in general?
 

**Anything else you would like to add:**

There has been a long discussion in the slack thread, xref: https://kubernetes.slack.com/archives/C8TSNPY4T/p1655805341632159

/kind feature

**Comment by furkatgofurov7** [comment]
Based on the slack conversation we had, this is a small summary of all points we had:

- there is an instruction on upstream k8s docs how to handle this: 
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version and whole project to ensure storage version for all CRDs:
https://github.com/kubernetes-sigs/kube-storage-version-migrator

Seems, the way we could  tackle this as of now is:

1. Run the [storage Version migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator)

1. Remove the old version from the CustomResourceDefinition status.storedVersions field.

based on the https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#previous-storage-versions

- Also, it was noted that this is a generic Kubernetes problem with migrations. API server does not migrate objects in bulk for reasons, but everything will be migrated over time. But if we upgrade again in a fast sequence before that happen we would have problems. In most use cases maybe just going manually over the CRDs, checking for old storage versions and then reapplying with the current version might be safer. However, to have a better user experience and let the clusterctl handle it, we could consider improving some parts of clusterctl (e.g. drop old unused versions from the CRDs before we update them)

- There were/are multiple upstream KEPs, 
https://github.com/kubernetes/enhancements/pull/3185 
should be the latest one (prior ones, i.e https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers was taken up in PRs [https://github.com/kubernetes/kubernetes/pull/96403 & ](https://github.com/kubernetes/kubernetes/pull/96403)[https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/85 ](https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/85) but those PRs stopped receiving updates and were marked as stale in mid 2021)


/cc @sbueringer @fabriziopandini @lentzi90 @chrischdi

**Comment by sbueringer** [comment]
Hey folks,
we've implemented a CRD migration in clusterctl upgrade in #6749 as we have to do it for cert-manager anyway.

I think it's a clean and straightforward solution for our issue and won't require a multi-month/year effort to get something implemented upstream in Kubernetes.

I looked over the KEPs and it seems like this will take a while:
* KEPs 2330, 2339, 2341, 2342, 2343
* [Replacement Draft](https://docs.google.com/document/d/11DN-dKcN7jEY1kvi8ZjlYjOCxK4FLKH_AsUL4TvFZqg/edit#heading=h.jtlgurs3tm7)
    * https://github.com/kubernetes/enhancements/issues/3166
    * https://github.com/kubernetes/enhancements/pull/3185

I would highly prefer a straightforward solution in clusterctl over modifying webhook and controllers in all providers to write all CRs on upgrades (this would also not work for cert-manager).

If there will be an upstream solution in either controller-runtime or Kubernetes over time we can of course deprecate our mechanism as it's then not necessary anymore
